### PR TITLE
Redesign handling of forward pointers

### PR DESCRIPTION
**Pull request by AlexeySotkin**
In some cases the current implementation doesn't create forward pointers
when they are required. New approach is to create forward pointers
during sorting the type instructions.

TODO:
- [x]  Add a LIT test

Signed-off-by: Alexey Sotkin <alexey.sotkin@intel.com>

### CODE REVIEW

**Comment by AlexeySotkin** [comment]
@MrSidims please take a look.

File: lib/SPIRV/libSPIRV/SPIRVModule.cpp [hunk 1]
```diff
@@ -1711,19 +1692,19 @@ class TopologicalSort {
   // the entry itslef.
   void visit(SPIRVEntry *E) {
     DFSState &State = EntryStateMap[E];
+    if (E->getOpCode() == OpTypePointer) {
+      SPIRVTypePointer *Ptr = static_cast<SPIRVTypePointer *>(E);
+      if (EntryStateMap[Ptr->getElementType()] == Discovered) {
```
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/SPIRVModule.cpp:1697]: May be it worth to add a comment, why do we insert OpTypeForwardPointer pointer if we found the same OpTypePointer in the module.

File: lib/SPIRV/libSPIRV/SPIRVModule.cpp [hunk 2]
```diff
@@ -1746,8 +1727,16 @@ class TopologicalSort {
   TopologicalSort(const SPIRVTypeVec &TypeVec,
                   const SPIRVConstantVector &ConstVec,
                   const SPIRVVariableVec &VariableVec,
-                  const SPIRVForwardPointerVec &ForwardPointerVec)
-      : ForwardPointerVec(ForwardPointerVec),
+                  SPIRVForwardPointerVec &ForwardPointerVec)
```
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/SPIRVModule.cpp:1730]: Why it's no longer a const?

**Review by MrSidims — COMMENTED** [review]
LGTM

**Review by AlexeySachkov — APPROVED** [review]
LGTM, but we indeed could add a comment about how we generate forward pointers as Dmitry suggested

File: lib/SPIRV/libSPIRV/SPIRVModule.cpp (hunk 2 above)
Comment (AlexeySachkov) [reply lib/SPIRV/libSPIRV/SPIRVModule.cpp:1730]: Because it is now modified in this function, see line 1754

Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRVModule.cpp:1730]: Indeed

File: lib/SPIRV/libSPIRV/SPIRVModule.cpp (hunk 1 above)
Comment (AlexeySotkin) [reply lib/SPIRV/libSPIRV/SPIRVModule.cpp:1697]: Done

**Review by MrSidims — APPROVED** [review]

File: test/transcoding/ForwardPtr.ll [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+%class.ArgFirst = type { void (%class.A addrspace(4)*)* }
+%structArgSec= type { %class.A addrspace(4)* }
+
+declare spir_func i1 @Caller(%struct.FuncArg addrspace(4)* ) align 2
+
+define spir_func void @MainFunc(%struct.FuncArg addrspace(4)* %context) {
+  entry:
+  %call = call spir_func zeroext i1 @Caller( %struct.FuncArg addrspace(4)* undef) 
+  ret void
+}
```
Comment (svenvh) [inline test/transcoding/ForwardPtr.ll:31]: Missing newline at end of file.

Comment (AlexeySotkin) [reply test/transcoding/ForwardPtr.ll]: Fixed.

**Pull request merged**